typed-protocols/src/Network/TypedProtocol/Channel.hs

```diff
@@ -158,6 +163,85 @@ createConnectedBufferedChannels sz = do
         recv   = atomically (Just <$> readTBQueue bufferRead)
 
 
+-- | Create a pair of channels that are connected via buffers with delays.
+--
+-- This is a crude approximation of asynchronous network connections where
+-- there is delay across the connection, and a limit on in-flight data.
+--
+-- The buffer size is bounded. It /blocks/ when 'send' would exceed the maximum
+-- buffer size. The size per channel element is provided by a function, so this
+-- can be size in bytes or a fixed size (e.g. 1 per element), so the max size
+-- should be interpreted accordingly.
+--
+-- The delays are modeled in a simplistic "GSV" style:
+--
+-- * The G is the minimum latency for a 0 sized message.
+-- * The S is the time per message size unit.
+-- * The V is the variance in latency, which is assumed to be uniform in the
+--   range @0..v@.
+--
+-- The sender is delayed by S. The receiver is delayed until the arrival time
+-- which is G + S + V.
```

Comment:
I'm not certain, but this feels somewhat incorrect. Something like "the receiver is delayed until either G + S + V, or by something more than that because the channel had to preserve order instead of respecting that particular scheduled arrival time."

This is the same thing I mentioned in the test code.